pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go

```diff
@@ -146,6 +147,20 @@ func (f *frontendSchedulerWorkers) getWorkersCount() int {
 	return len(f.workers)
 }
 
+// sendRequestCancel sends cancellation to the "already scheduled" frontendRequest.
+// It will make sure the frontend worker that is responsible for the `reqID`
+// receives the cancel signal.
+func (f *frontendSchedulerWorkers) sendRequestCancel(reqID uint64, cancelCh chan<- uint64) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
```

Comment:
is the lock just for f.workers ? if yes please then test it and release the lock before blocking on the send of the channel.

Reply:
Yea. in fact that was my initial fix using already existing [f.workers.getWorkersCount](https://github.com/grafana/loki/blob/3e3e23826b8ab1b149faa04387fce90ebdc79b47/pkg/lokifrontend/frontend/v2/frontend_scheduler_worker.go#L143). Just get `len(f.workers)` with locks and do the channel operation after that. But that may lead to a race right?

In fact it may leads to two problems.

say If I do like this (`getWorkersCount()` release lock already).

````
if f.workers.getWokersCount() > 0 {
            cancelCh <- f.requestID
}
````

1. There is a race between `if getWorkersCount() > 0` and `cancelCh <- f.requestID`.
2. If workers count > 0 during if, and it goes to 0 during `cancelCh <- f.requestID` then it goes for waiting it forever, because no one is listening on the other side of the channel (because no frontend workers)

It cannot go to blocking on send if we have non-zero frontend worker running. And both non-zero count check and send should happen with lock.

Reply:
Yeah that seems correct you are right !

Reply:
Also for other reviewer the lock is only held when adding/starting or removing workers. That should be fast and non blocking.

Reply:
That seems correct yep.

Reply:
@colega you are absolutely right, `cancelCh` that is part of frontendReq belongs to single woker. And hard to assume if that worker will ever get that.

Smart thinking :)